## Day 49 Lecture 2 Assignment

In this assignment, we will apply mean shift clustering to a dataset containing the results of a survey on financial wellbeing.

In [1]:
%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.cluster import MeanShift
from sklearn.preprocessing import StandardScaler

This dataset contains the results of a survey on a financial wellbeing conducted by the US Consumer Finance Protection Bureau that were published in October 2017. This dataset has a large number of columns, most of which correspond to specific questions on the survey. These codebook for translating the column names to questions can be found here:

https://s3.amazonaws.com/files.consumerfinance.gov/f/documents/cfpb_nfwbs-puf-codebook.pdf

Load the dataset.

In [2]:
# answer goes here
finance = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/Data%20Sets%20Clustering/financial_wellbeing.csv')
print(finance.shape)
finance.head()

(6394, 217)


PUF_ID  sample  fpl  SWB_1  ...  PPT1317  PPT18OV  PCTLT200FPL   finalwt
0   10350       2    3      5  ...        0        1            0  0.367292
1    7740       1    3      6  ...        0        2            0  1.327561
2   13699       1    3      4  ...        1        2            1  0.835156
3    7267       1    3      6  ...        0        1            0  1.410871
4    7375       1    3      4  ...        0        4            1  4.260668

[5 rows x 217 columns]

While the survey questions have the potential for interesting cluster analysis, we will stick to the "score" columns to avoid clustering in an unreasonably high-dimensional space. The columns we are interested in all have "score" in their names; identify and isolate these columns. (There should be 4 in total.)

In [3]:
# answer goes here
#filtering for score columns
score = finance.filter(like='score', axis=1)
print(score.shape)
score.head()

(6394, 4)


FWBscore  FSscore  LMscore  KHscore
0        55       44        3    1.267
1        51       43        3   -0.570
2        49       42        3   -0.188
3        49       42        2   -1.485
4        49       42        1   -1.900

Standardize the features in your dataset using scikit-learn's StandardScaler, which will set the mean of each feature to 0 and the variance to 1.

In [4]:
# answer goes here
#scaling features
scale = StandardScaler()
X_scale = pd.DataFrame(scale.fit_transform(score), columns=score.columns)
X_scale

FWBscore   FSscore   LMscore   KHscore
0    -0.073062 -0.530906  0.653830  1.624716
1    -0.355677 -0.609920  0.653830 -0.629626
2    -0.496984 -0.688935  0.653830 -0.160841
3    -0.496984 -0.688935 -0.670399 -1.752502
4    -0.496984 -0.688935 -1.994628 -2.261785
...        ...       ...       ...       ...
6389  0.350859 -0.293863  0.653830  1.624716
6390  0.209552  0.654309 -1.994628 -1.421162
6391  0.209552  0.022194 -0.670399 -1.421162
6392 -0.708944  0.259237 -0.670399 -1.421162
6393 -0.496984 -0.688935 -0.670399 -1.421162

[6394 rows x 4 columns]

Run mean shift clustering on the scores in the survey dataset using the default bandwidth. Then answer the following by printing or typing as appropriate:

- How many clusters are produced? 
- What are the cluster centers?
- How many responses are assigned to each cluster?
- Are these results reasonable? If not, what changes should we make?

In [6]:
# answer goes here
#use default bandwidth
ms = MeanShift(bandwidth=None)
score['MS_bN'] = ms.fit_predict(X_scale)

print(
    f'There are {score.MS_bN.nunique()} clusters\n'
    f'The amount of values in each cluster are:\n {score.MS_bN.value_counts()}\n'
    )

#cluster centers
clusters = scale.inverse_transform(ms.cluster_centers_)
clusters = pd.DataFrame(clusters, columns=X_scale.columns)
print('Cluster centers for bandwith=None')
clusters.style.background_gradient()

There are 2 clusters
The amount of values in each cluster are:
 0    6326
1      68
Name: MS_bN, dtype: int64

Cluster centers for bandwith=None


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


It seems that the bandwidth when set to None is too high of a bandwith because there are a disproportionate amount of data points in the 2 clusters (6326 in one and only 68 in the other). When you compare the scores, especially the FWBscore and FSscore, it shows a hugh difference in the cluster centers. A way to possibly fix the results is to set a smaller bandwidth and in turn that would create more clusters.

Try changing the appropriate parameters of the mean shift algorithm to achieve a better clustering result. Answer all of the same questions from the previous clustering step.

In [8]:
# answer goes here
#experimenting with bandwidth
ms = MeanShift(bandwidth=1)
score['MS_b1'] = ms.fit_predict(X_scale)

#results
print(
    f'There are {score.MS_b1.nunique()} clusters\n'
    f'The amount of values in each cluster are:\n {score.MS_b1.value_counts()}\n'
    )

#cluster centers
clusters = scale.inverse_transform(ms.cluster_centers_)
clusters = pd.DataFrame(clusters, columns=X_scale.columns)
clusters.style.background_gradient()

There are 41 clusters
The amount of values in each cluster are:
 0     3215
1     1353
2      441
4      229
9      132
7      130
25     119
16     117
15     100
3       85
5       74
17      63
14      54
6       39
13      30
10      25
11      24
24      19
32      18
8       14
22      13
35      12
36      12
37      11
27      11
34      10
38       6
30       5
40       5
12       5
33       4
29       4
18       3
20       2
31       2
21       2
19       2
23       1
28       1
26       1
39       1
Name: MS_b1, dtype: int64



/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [9]:
ms = MeanShift(bandwidth=2)
score['MS_b2'] = ms.fit_predict(X_scale)

#results
print(
    f'There are {score.MS_b2.nunique()} clusters\n'
    f'The amount of values in each cluster are:\n {score.MS_b2.value_counts()}\n'
    )

#cluster centers
clusters = scale.inverse_transform(ms.cluster_centers_)
clusters = pd.DataFrame(clusters, columns=X_scale.columns)
clusters.style.background_gradient()

There are 2 clusters
The amount of values in each cluster are:
 0    6326
1      68
Name: MS_b2, dtype: int64



/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [10]:
ms = MeanShift(bandwidth=1.5)
score['MS_b15'] = ms.fit_predict(X_scale)

#results
print(
    f'There are {score.MS_b15.nunique()} clusters\n'
    f'The amount of values in each cluster are:\n {score.MS_b15.value_counts()}\n'
    )

#cluster centers
clusters = scale.inverse_transform(ms.cluster_centers_)
clusters = pd.DataFrame(clusters, columns=X_scale.columns)
clusters.style.background_gradient()

There are 8 clusters
The amount of values in each cluster are:
 0    4096
1    2156
5      50
4      24
2      23
7      22
6      12
3      11
Name: MS_b15, dtype: int64



/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [11]:
ms = MeanShift(bandwidth=1.7)
score['MS_b17'] = ms.fit_predict(X_scale)

#results
print(
    f'There are {score.MS_b17.nunique()} clusters\n'
    f'The amount of values in each cluster are:\n {score.MS_b17.value_counts()}\n'
    )

#cluster centers
clusters = scale.inverse_transform(ms.cluster_centers_)
clusters = pd.DataFrame(clusters, columns=X_scale.columns)
clusters.style.background_gradient()

There are 4 clusters
The amount of values in each cluster are:
 0    5946
1     409
3      20
2      19
Name: MS_b17, dtype: int64



/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [12]:
ms = MeanShift(bandwidth=0.7)
score['MS_b07'] = ms.fit_predict(X_scale)

#results
print(
    f'There are {score.MS_b07.nunique()} clusters\n'
    f'The amount of values in each cluster are:\n {score.MS_b07.value_counts()}\n'
    )

#cluster centers
clusters = scale.inverse_transform(ms.cluster_centers_)
clusters = pd.DataFrame(clusters, columns=X_scale.columns)
clusters.style.background_gradient()

There are 146 clusters
The amount of values in each cluster are:
 0      1514
1      1390
2       762
3       236
9       169
       ... 
94        1
96        1
92        1
81        1
143       1
Name: MS_b07, Length: 146, dtype: int64



/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [13]:
ms = MeanShift(bandwidth=0.9)
score['MS_b09'] = ms.fit_predict(X_scale)

#results
print(
    f'There are {score.MS_b09.nunique()} clusters\n'
    f'The amount of values in each cluster are:\n {score.MS_b09.value_counts()}\n'
    )

#cluster centers
clusters = scale.inverse_transform(ms.cluster_centers_)
clusters = pd.DataFrame(clusters, columns=X_scale.columns)
clusters.style.background_gradient()

There are 59 clusters
The amount of values in each cluster are:
 0     3104
1     1339
2      395
8      163
4      138
12     115
33     109
24     109
5      103
13      90
14      83
3       70
54      70
23      51
16      37
6       32
20      30
19      28
7       24
53      24
10      21
43      20
32      19
55      19
9       18
56      17
37      12
47      11
35      11
46      11
11      10
17      10
50       9
38       8
49       8
21       7
52       7
18       7
22       6
40       5
51       5
15       5
39       4
30       4
25       4
28       3
26       3
29       2
36       2
27       2
42       2
45       1
48       1
44       1
34       1
58       1
31       1
41       1
57       1
Name: MS_b09, dtype: int64



/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


I would say that the bandwidth for the MeanShift should be between 0.9 and 1.4 as that would illustrate the different cluster centers better and at the same not form centers around individual data points.